### DISCUSSION TITLE
cannot get GPIO input to work with MQTT - SOLVED

### DISCUSSION THREAD

**Discussion by knx-gvs** [arendst/Tasmota#12079] [Q&A] [answered]
Hi - newbie.

I want to use MQTT to see the status of an input (following switch) on my sonoff SV flashed with tasmota. I have already setup as follows:

1) wifi network is working
2) configuration GPIO14 set as switch
3) switchmode 1 - done
4) SetOption114 1 - done
5) MQTT working - I can switch power relays 1 and 2 from my node red setup
6) topic is gate

when I then type stat/gate/switch in the console I get 
19:20:19.440 CMD: stat/gate/switch
19:20:19.448 MQT: stat/gate/RESULT = {"Command":"Unknown"}

when I then type stat/gate/sensor in the console I get 
19:20:50.465 CMD: stat/gate/sensor
19:20:50.473 MQT: stat/gate/RESULT = {"Command":"Error"}

toggling the input itself gives the following in the console so the switch is working - but I don't know how to check the stat via mqtt command.
19:26:18.416 MQT: stat/gate/RESULT = {"Switch1":{"Action":"ON"}}
19:26:26.231 MQT: stat/gate/RESULT = {"Switch1":{"Action":"OFF"}}

I have no idea what to set my node-red mqtt input to view what the switch is doing - I'd like to use this input to connect a reed switch to it to monitor whether the gate is open or closed. I can already open and close the gate from my homekit - this is working fine. I just cant get the input working to update the status internally. what am I doing wrong?? have a missed other settings? I've also tried setting up the configure template to reflect the GPIOs but this also did not work - I'm actually not sure how the template differs from the module configuration menu and how these tie in together if someone can enlighten me. 

thanks in advance

**Comment by barbudor** [comment, marked as answer]
you don't have to " to check the stat via mqtt command."
tasmota publish the state so you just have to subscribe to it

`Switchmode1 15` is an alternative to have the switch state automatically published in the SENSOR message.
In NodeRed, subscribe to the SENSOR message
And you can use the `SwitchText1 Gate` command to rename `Switch1` to `Gate` in the SENSOR message

But you can also still use SetOption114. You just have to subscribe to the right topic.
SetOption4 May help you get a more easier to use topic to subscribe to

**Reply by sfromis** [reply]
To have commands in the console, you should not use MQTT topics, just the command name, which could be Power and many others. As barbudor describes, you should generally not ask for switch states, as Tasmota can publish changes arriving on your subscription as it happens. It is still possible to get switch states using a command like `Status 10` in the console, or by publishing such a command via MQTT using the appropriate topic structure.

**Reply by knx-gvs** [reply]
Hi Guys - ok I made some progress thanks for the help. yes, this is how I understand the mqtt - that changing the switch publishes the topic which must then be subscribed to, I was trying to use the console to try to figure out what the topic is I'm supposed to subscribe. thanks to https://www.youtube.com/watch?v=qZqmkhpJ55A for his video on a door switch. He explains how to change the switchtopic to something useable - which I called gate_open -> then I subscribed to the topic cmnd/gate_open/POWER1 in node red and that is working now. I also see the POWER1 being published along with the switch action result in the console. that wasn't there before. something I'm still confused about is when to use tele, cmnd and stat when subscribing or publishing (yes read the documentation - still not certain). I see now that the cmnd/gate_open/power1 works but I honestly don't know why. none the less - it would have seemed more logical (for me at least) if I could simply subscribe to cmnd/gate/switch<x> - i.e. changing then switch publishes a message saying its on or off and the simply subscribe to switch - don't know why I must subscribe to power<x> as for me this was analogous to subscribing to the relay state. so ok - now if I understand this correctly I am subscribing to cmnd/new switch topic name/power<x>. regarding the documentation on switchmode1 15 -> mode 15 is not documented @ https://tasmota.github.io/docs/Commands/ but is under @ https://tasmota.github.io/docs/Buttons-and-Switches/#switchmode hence I did not see that one the first round. handy thanks.

**Reply by sfromis** [reply]
Using a cmnd/ topic to publish status messages (which should be stat/) is really a misuse of the prefix. With Tasmota, cmnd/ is only meant for sending commands to a device. stat/ is generally meant for the device to publish status info about what happens, and tele/ is telemetry published every TelePeriod, not meant to drive automations. While this separation is the deeper meaning, in some cases, tele/ and stat/ are not clearly sorted out in a consistent way.

What I'd suggest doing is to read the MQT: messages in the Tasmota console where you can see all messages Tasmota publishes when. This is an excellent basis for deciding what is useful for your automations.

The thing about status messages unfortunately being misleadingly and confusingly published to the command topic should really be avoided, but some time ago, it was also being used in an example in Tasmota docs.

**Reply by knx-gvs** [reply]
excellent thanks that makes sense - so use stat instead of cmnd when subscribing. I'll try that tonight. If everything is working then I can close this topic. just one more question about that power<x> instead of switch (or sensor for that matter when using switchmode 15) - why power? as I said that was analogous to the relay state to me - only difference now is that it's been disconnected from the relay via setoption114.

**Reply by sfromis** [reply]
When using "power", it is not instead of switch, but as a reflection of the power status being changed, possibly due to switch actions (and other things too). Not sure what you're specifically referring to, but SwitchMode 15 (which is similar to SetOption114) is specifically described as not controlling power outputs, and that the sensor payload published includes names like "Switch1".

**Reply by knx-gvs** [reply]
ok thanks everyone. just one or two feedback items: I tried using stat instead of the cmnd command as suggested but didn't work. so I understand the abuse situation but I'll have to go with that for now at least. 

This topic can be closed.
